modules/engine-prototype/src/main/scala/engine/package.scala

```diff
@@ -0,0 +1,46 @@
+import scala.concurrent.ExecutionContext
+
+import cats.Apply
+import cats.implicits._
+import cats.effect.Effect
+
+import fs2.Stream
+import fs2.async
+import fs2.async.mutable.Queue
+
+package object engine {
+
+  def run[F[_]](in: Queue[F, Event])(st0: Sequence.State)(implicit F: Effect[F], ec: ExecutionContext): Stream[F, Sequence.State] =
+    Sequence.State.Mutable.stream[F](st0: Sequence.State)(m => forever(reader(in)(m)))
+
+  def reader[F[_]](in: Queue[F, Event])(m: Sequence.State.Mutable[F])(implicit F: Effect[F], ec: ExecutionContext): F[Unit] =
+    in.dequeue1.flatMap {
+      case Event.Start => m.withState(st0 =>
+        st0.status match {
+          case Status.Waiting => {
+            val st1 = Sequence.State.status.set(Status.Running)(st0)
+            st1 match {
+              // Execute only when the State is in ready position. If it's not
+              // it's either already executing or it doesn't make sense to execute,
+              // like when it's done.
+              case Sequence.State.ready(p) => async.fork(p.execute(m)) *> F.pure(st1)
```

Comment:
Before returning the *reaction* to the start event we launch another async `Task` (`Effect` or whatever) that takes care of executing and is not responsible of handling user input. The current threads goes back to listening incoming input events.